ckanext/harvest/logic/action/harvest_additions.sh

```diff
@@ -0,0 +1,111 @@
+#!/bin/bash
+# Purpose: Read CSV File and create harvest sources
+# Author: James Brown
+# ------------------------------------------
+
+# Display Help
+Help()
+{
+    echo "Adds harvest sources from a CSV file"
+    echo
+    echo "Syntax: harvest_additions [-f|h|k|o]"
+    echo "options:"
+    echo "[required] -f, --file             Path to CSV file"
+    echo "-h, --help                        Print this Help"
+    echo "[required] -k, --api-key          CKAN API key"
+    echo "[required] -o, --organization     CKAN organization name"
+    echo "-u, --url                         Base URL for CKAN instance"
+    echo
+    exit 0
+}
+
+#Default value for testing
+URL='localhost:8080'
+
+#Argument parsing
+POSITIONAL=()
+while [[ $# -gt 0 ]]
+do
+key="$1"
+
+case $key in
+    -f|--file)
+    FILE="$2"
+    shift # past argument
+    shift # past value
+    ;;
+    -h|--help)
+    Help
+    shift # past argument
+    ;;
+    -k|--api-key)
+    API_KEY="$2"
+    shift # past argument
+    shift # past value
+    ;;
+    -o|--organization)
+    ORGANIZATION="$2"
+    shift # past argument
+    shift # past value
+    ;;
+    -u|--url)
+    URL="$2"
+    shift # past argument
+    shift # past value
+    ;;
+    --default)
+    DEFAULT=YES
+    shift # past argument
+    ;;
+    *)    # unknown option
+    POSITIONAL+=("$1") # save it in an array for later
+    shift # past argument
+    ;;
+esac
+done
+set -- "${POSITIONAL[@]}" # restore positional parameters
+
+# If arguments are not given/are not correct, exit
+if [ ! -f "$FILE" ] || [ -z "$ORGANIZATION" ] || [ -z "$API_KEY" ]; then
+    echo "Please specify the organization, API key, and file path. Use --help for more information."
+    exit 99
+fi
+
+generate_harvest_data()
+{
+    cat <<EOF
+{
+    "title": "$title",
+    "name": "$name",
+    "url": "$waf_url",
+    "owner_org": "$ORGANIZATION",
+    "source_type": "waf",
+    "config": "{\"private_datasets\": false}",
+    "frequency": "$frequency"
+}
+EOF
+}
+
+OLDIFS=$IFS
+IFS=','
```

Comment:
As soon as we're dealing with CSV, I would not use shell because of the quoting issues. I would use python since it's csv module handles all kinds of quoting options for parsing/writing CSVs.

Reply:
Yeah, this probably could have been with Python. I just didn't want to write in Python2 and use deprecated libraries and then have to upgrade in the future, or use Python3 and it wouldn't work on the system themselves.